### DISCUSSION TITLE
Session expires immediatly with remote infinispan cluster and JDBC persistence

### DISCUSSION THREAD

**Discussion by daviddelannoy** [keycloak/keycloak#10577] [Q&A]
Hi,

We are facing a strange bug with a remote Infinispan cluster **with JDBC persistence** and Keycloak (wildfly based).
I agree that it is quite an unusual set-up but we have to make it work...

Keycloak 17.0.0 (legacy, wildfly) (will be migrated to Quarkus in a few months)
Infinispan 13.0.6.Final
PostgreSQL 14.X

**Note : this bug does not occur with a remote Infinispan cluster without JDBC persistence.**

**Scenario :**

https://github.com/olivierboudet/keycloak-infinispan-jdbc

docker-compose up 

- Login to [http://localhost:8080/js-console](url) with test/test
- Look at new Session timestamp in database (correct = Current time + SSO session Max (10 hours)): 

![bdd1](https://user-images.githubusercontent.com/16318239/156616227-7cdf1b07-3631-4c4c-bcc7-7859a50ad9ab.png)

- Then go on another application, click on "**Account**" button in js-console, we are already logged in, OK.
- Look at existing session timestamp in database. It has been updated with wrong timestamp => 01/01/1970 + SSO session Max (10 hours) instead of Current timestamp + SSO session Max (10 hours)
- => session quickly removed, user logged out

![image](https://user-images.githubusercontent.com/16318239/156616377-28cfd917-c8d5-421f-9e50-3bb404f1764d.png)

We have looked at theses classes and put them in DEBUG but were not able to identify the root cause : 

org.keycloak.models.sessions.infinispan.remotestore.RemoteCacheSessionListener
org.keycloak.models.sessions.infinispan.remotestore.RemoteCacheInvoker
org.keycloak.models.sessions.infinispan.InfinispanUserSessionProvider

If anyone can help us on this problem ... 
Is there a big mistake in our configuration ? we simplified it a bit for the example (remove security, single Keycloak/ISPN instance, ...)

thanks

**Comment by daviddelannoy** [comment]
Hi,

litlle up, in case someone has ever tried ISPN JDBC persistence ... 

thanks

**Comment by thomasdarimont** [comment]
@daviddelannoy thansk for the useful reproducer!

I also tested this with a custom docker-image from the latest main branch (18.0.0-SNAPSHOT) and there I see the same problem.

I think the problem is here: https://github.com/keycloak/keycloak/blob/main/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remotestore/RemoteCacheInvoker.java#L132
```java
case REPLACE:
    replace(topology, remoteCache, task.getLifespanMs(), maxIdleMs, key, task);
    break;
```

`task.getLifespanMs()` is exactly the too small value that gets written to the remote store... if one passes `-1` instead of `task.getLifespanMs()` I get the "correct"? new value written down to the timestamp column.

I just gave this a spin - the following PoC (Patch included) works for me:
https://github.com/thomasdarimont/keycloak-infinispan-jdbc/tree/poc

I don't know yet whether this would be a proper fix but other folks like @mposolda know the infinispan internals better :)

**Reply by daviddelannoy** [reply]
Hi @thomasdarimont 

Thanks a lot for your answer ! Indeed, there may be something to fix in method `InfinispanUtil.toHotrodTimeMs()` to handle properly the replace case.

You may have a cross-question from @olivierboudet on another discussion as we are currently trying to consider the best option regarding session persistence : 

- direct jdbc-store on keycloak embeded ISPN distributed-caches (session, clientSessions)
- external ISPN cluster with remote JDBC store (like [this example](https://github.com/thomasdarimont/keycloak-project-example/tree/main/deployments/local/clusterx/haproxy-external-ispn) that you've updated recently with digest authentication for Hotrod)

We will do some benchmarks with keycloak-benchmark project but what will be your advice with the requirement "_user sessions must be persisted in database_" ? 
Especially about Quarkus future releases  and support of remote ISPN cluster.
Performances will also be impacted for sure with such configuration, but it's a trade-off for persistence

thanks again

**Reply by daviddelannoy** [reply]
Hi @mposolda @thomasdarimont 

Should I reference this discussion in a new issue ? 

Thanks

**Reply by thomasdarimont** [reply]
@daviddelannoy sure :)

**Reply by sstepien** [reply]
Hi @thomasdarimont 

We also have this requirement: _"user sessions must be persisted in database"_. We migrated from Keycloak 10.0.2 to KeycloakX 17.0.0 but it still it is unclear for us how to achieve it. Ideally we would like to use direct jdbc-store on keycloak embeded ISPN distributed-caches (session, clientSessions).
Are there any plans to add support for this or what are the alternatives?

Thanks

**Reply by thomasdarimont** [reply]
@daviddelannoy I jsut gave this another spin and tested this with Keycloak 17.0.1
The patch seems to do the trick:
```java
            case REPLACE:
                // BEGIN-PATCH
                long lifespanPatch = Time.currentTimeMillis() + task.getLifespanMs();
                long maxIdlePatch = Time.currentTimeMillis() + maxIdleMs;
                replace(topology, remoteCache, lifespanPatch, maxIdlePatch, key, task);
//                replace(topology, remoteCache, -1, maxIdleMs, key, task);
                // END-PATCH
```
See: https://github.com/thomasdarimont/keycloak-infinispan-jdbc/commit/49efacdd3f6db9f0ad36ddd83befd065884a19c1

**Reply by daviddelannoy** [reply]
Hi [@thomasdarimont](https://github.com/thomasdarimont)

Yes thanks, it seems to be OK like that. I created an issue from this discussion but I think it is still in triage : [10755](https://github.com/keycloak/keycloak/issues/10755)

Just like [@sstepien](https://github.com/sstepien) in message above, in the meantime we tested direct JDBC store but faced another problem on which you also have proposed a patch (see [10803](https://github.com/keycloak/keycloak/issues/10803) (thanks for that ).

So right now, two options are available to us, the two of them require their own patch and documentation is not very clear about what should be the recommended one, especially with Quarkus (except the : do not persist user and client sessions ^^)

Any input on this subject will help us (and others) to choose the most reliable solution
